chapter_optimization/minibatch-sgd_vn.md

```diff
@@ -511,9 +508,8 @@ Minibatch SGD is able to trade-off the convergence speed and computation efficie
 A minibatch size 10 is more efficient than SGD; a minibatch size 100 even outperforms GD in terms of runtime.
 -->
 
-Cuối cùng, chúng ta so sánh tương quan thời gian và giá trị mất mát trong bốn thí nghiệm trên.
-Như có thể thấy, mặc dù SGD hội tụ nhanh hơn GD về số mẫu được xử lý,
-nhưng SGD tốn nhiều thời gian hơn để đạt được cùng giá trị mất mát như GD vì thuật toán này tính toán gradient trên từng mẫu một.
+Cuối cùng, chúng ta so sánh tương quan thời gian và giá trị hàm mất mát trong bốn thí nghiệm trên.
+Có thể thấy, SGD tốn nhiều thời gian hơn để đạt được cùng giá trị mất mát như GD vì thuật toán này tính toán gradient trên từng mẫu một. <!--câu `mặc dù SGD hội tụ nhanh hơn GD về số mẫu được xử lý` cứ sai sai.-->
```

Comment:
Cái này đúng mà. SGD hội tụ nhanh hơn nếu so sánh về số mẫu.  Còn về thời gian xử lý thì chậm hơn. (Có thể tham khảo biểu đồ của chương này để hiểu rõ hơn)

Reply:
mình không hiểu `nhanh hơn về số mẫu` là sao? Khi so sánh nhanh chậm hoặc là fix cùng số mẫu, đo thời gian hoặc là fix thời gian, đo số mẫu xử lý. Dù thế nào thì GD vẫn nhanh hơn chứ?

Reply:
Với GD, loss của nó theo epoch sẽ là:
![image](https://user-images.githubusercontent.com/23548268/87228464-a5766a80-c3cb-11ea-8b11-cd854e0d41c6.png)
Với SGD thì là:
![image](https://user-images.githubusercontent.com/23548268/87228488-c8a11a00-c3cb-11ea-9dda-397ebd79507d.png)
Có thể thấy nếu nhìn số epoch thì SGD hội tụ nhanh hơn so với GD do một epoch SGD được cập nhật tới 1500 lần với hướng gradient có thể không hoàn toàn chính xác so với GD nhưng cũng đủ để dẫn nó về điểm hội tụ. Mặc khác, GD cho hướng chuẩn hơn nhưng nó chỉ cập nhật được 1 lần trong 1 epoch => hội tụ chậm hơn SGD

Nhưng về mặt thời gian xử lý, SGD tốn nhiều thời gian hơn vì không sử dụng tối ưu vector hóa. 
> GD loss: 0.254, **0.076** sec/epoch

> SGD loss: 0.243, **0.428** sec/epoch

Reply:
hm trong hình GD thì 2 epoch đầu không được plot, SGD thì chỉ plot 2 epoch đầu, nên mình nghĩ không khẳng định được là SGD cần ít epoch để hội tụ hơn.
> Mặc khác, GD cho hướng chuẩn hơn nhưng nó chỉ cập nhật được 1 lần trong 1 epoch => hội tụ chậm hơn SGD

Ý này (tạm coi là đúng) thì mình cũng chưa thấy liên quan gì đến `hội tụ nhanh hơn về số mẫu được xử lý`.

Reply:
Câu gốc là `SGD converges faster than GD in terms of number of examples processed` thì mình hiểu tác giả muốn nói đến với cùng một số lượng mẫu, SGD tạo được nhiều progress hơn GD. Mà progress ở đây là việc hội tụ. Nhanh ở đây không theo đơn vị thời gian mà theo đơn vị số mẫu xử lý được

Reply:
```suggestion
Có thể thấy, dù hội tụ nhanh hơn GD về số mẫu được xử lý, nhưng SGD tốn nhiều thời gian hơn để đạt được cùng giá trị mất mát như GD vì thuật toán này tính toán gradient trên từng mẫu một.
```
mình vẫn thấy ý đó gây confused rất nhiều, cơ mà nếu mn thấy ổn thì cứ để vậy.

Reply:
Yeah, mình cũng hiểu vì cách tác giả plot khá khó để nói được điều đó